### PR TITLE
Use Option for ExecutorService in AsyncHttpClient.apply

### PR DESCRIPTION
**Pull request by ceedubs**
Based on [this discussion](https://gitter.im/http4s/http4s?at=56e9c7c86fde057c26876c95).

I think that if we do this we should go further and do a similar thing
in other places (the Blaze client, etc). We can probably completely
remove the `DefaultExecutorService` marker. I wanted to make sure I was
on the right track before changing everything.

### CODE REVIEW

File: async-http-client/src/main/scala/org/http4s/client/asynchttpclient/AsyncHttpClient.scala [hunk 1]
```diff
@@ -42,17 +42,18 @@ object AsyncHttpClient {
...
   def apply(config: AsyncHttpClientConfig = defaultConfig,
             bufferSize: Int = 8,
-            executorService: ExecutorService = DefaultExecutor.newClientDefaultExecutorService("async-http-client-response")): Client = {
+            customExecutor: Option[ExecutorService] = None): Client = {
     val client = new DefaultAsyncHttpClient(config)
-    val close = executorService match {
-      case es: DefaultExecutorService =>
+    val executorService = customExecutor.getOrElse(DefaultExecutor.newClientDefaultExecutorService("async-http-client-response"))
+    val close =
+      if (customExecutor.isDefined)
```
Comment (ceedubs) [inline async-http-client/src/main/scala/org/http4s/client/asynchttpclient/AsyncHttpClient.scala]: Doing this made me realize I have no idea how to format a multi-line if/else in Scala. I guess I hardly ever use if/else!

**Comment by rossabaker** [comment]
So far, so good.

**Comment by ceedubs** [comment]
Thanks @rossabaker. I should be able to follow up with the rest of the changes within the next few days.

**Comment by rossabaker** [comment]
This pattern seems nice for AsyncHttpClient, which needs to be closed regardless of the custom executor.

I'm currently poking at OAuth2 support, and that needs an executor, which is the only thing that needs to be closed.  So I have to provide a close for that in case of a default executor, but it's a no-op when a custom executor is passed.  That seems awkward.

**Comment by ceedubs** [comment]
Hmm this isn't so straightforward when I try to take the same approach with `BlazeClient`. It becomes trickier because `Http1Support`, `PooledHttp1Client`, and `BlazeClient` all want to try to share the same executor. `BlazeClient` wants to set up its `shutdown` `Task`, but never exposes the underlying executor such that `Http1Support` and `PooledHttp1Client` could reuse it.

One solution that comes to mind is that maybe we should be able to pass in a `Task[Unit]` to run on shutdown? Or maybe `ExecutorService => Task[Unit]`? Open to ideas here.

**Comment by bryce-anderson** [comment]
I'd say pass a Task to the BlazeClient constructor: that's pretty much what
is happening in there anyway.

It feels like a bit of this noise is caused by the number of places
involved in constructing a client in blaze. I think the PooledClient and
SimpleClient package objects should go and the constructors should go into
BlazeClient.

**Pull request closed without merging**